### PR TITLE
Add activity reference material v1

### PR DESCRIPTION
**Pull request by tonyshumlh**

### CODE REVIEW

**Review by kvarada — CHANGES_REQUESTED** [review]
Thank you for doing this @tonyshumlh ! 

I think it would be more useful and productive for our audience if we provide them with several publicly available, messy datasets alongside specific objectives. We can then ask them to choose one dataset and address the following questions:

- Is the provided dataset appropriate for the specified objective? What type of data would ideally solve your problem or research question? Are there better-suited datasets available for this objective?
- Clearly define the expected input and the 'ideal' output. Determine if machine learning is the appropriate method for addressing this problem.
- If machine learning is deemed suitable, what should the model aim to achieve? How would you measure the model's performance?
- How would a human tackle this issue? Can you propose any heuristic methods to solve this problem?
- What are the major steps required to resolve this problem?
- Draw a diagram that illustrates the input, output, and key stages of the problem-solving process.
- Which type of machine learning would be best suited for your problem?
- What specific machine learning technique would be most effective for this problem?

It would be great to show a worked out example for them so that they have something to refer to. For example, you can take the Credit Card Fraud Detection dataset (https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud) or the Adult Census Dataset (https://archive.ics.uci.edu/dataset/2/adult) as an example and try to answer the questions above.

**Comment by yuliaUU** [comment]
@kvarada  thank you! Are we still doing 2 sections in teh afternoon section? for those who do know how to code and those who does not? Or we merge everyone together?

**Comment by kvarada** [comment]
@yuliaUU Based on the survey results, I think it makes sense to have just one section focused on ML problem framing. That seems like the most useful approach for this audience. Some people will bring their own problems, while others will work on the datasets we provide. We’ll offer guiding questions (like the ones mentioned above) to help with ML problem framing. What do you think?

**Comment by kvarada** [comment]
@tonyshumlh and @yuliaUU 
FYI: The above write-up for the lab session will go [here](https://github.com/kvarada/DSI-ML-workshop-2024/blob/main/website/workshop-10.qmd) in the GitHub repo and [here](https://kvarada.github.io/DSI-ML-workshop-2024/workshop-10.html) on our website.

**Comment by yuliaUU** [comment]
@kvarada yes, better to keep things simple! i like the idea that students can bring their dataset and get help with discussing what can work best for their data

**Comment by tonyshumlh** [comment]
---
title: "Topics and Data Sources for Machine Learning applications"
format: 
    revealjs:
      smaller: true
      center: true
---

## Potential Topics for Machine Learning Projects

Below are some sectors and topics of real-world issues that could possibly be solved using Machine Learning

### Business

#### Finance

-   forecast the direction of stock market index price change based on financial indicators
-   detect potentially fraudulent transactions of credit cards

#### Marketing

-   estimate the effectiveness of a marketing campaign by predicting customer action (click/not click, purchase/not purchase, etc.)
-   predict whether a customer will churn after a certain period of time
-   customer segmentation based on user behaviors on social media platform for targeted marketing

### Science & Engineering

#### Biology

-   identify and classify genomic sequences from DNA samples
-   identify and detect malignant tumors cell based on histology imaging

#### Structural Engineering

-   predict the risk of building/facility collapse when earthquake occurs based on key indicators and building images

### Social Science

#### Psychology

-   estimate how likely a person has autism or other developmental disorder based on personal traits and history

#### Public Health

-   forecast the CO2 emission of a country based on the country's statistics (demographics, economic activity, etc.)
-   predict the risk of cardiovascular disease of an individual based on key health indicators


## Common Open Data Source for Machine Learning Projects

Below are some open data sources that could be used for Machine Learning projects

### General

-   [UCI Machine Learning Repository](https://archive.ics.uci.edu/datasets)
-   [Kaggle](https://www.kaggle.com/datasets)

### Business

-   [Federal Reserve Bank of St. Louis](https://fred.stlouisfed.org/)

### Science & Engineering

-   [National Library of Medicine](https://pubchem.ncbi.nlm.nih.gov/)

### Social Science

-   [Gapminder](https://www.gapminder.org/data/)
-   [The Humanitarian Data Exchange (HDX)](https://data.humdata.org/dataset)
-   [Statistics Canada](https://www150.statcan.gc.ca/n1/en/type/data?MM=1#tables)

## ML Problem Framing Example 

### Data 
- Credit Card Fraud Detection from [Kaggle](https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud)

### Framing
1. Is the provided dataset appropriate for the specified objective? What type of data would ideally solve your problem or research question? Are there better-suited datasets available for this objective?
- The ideal dataset should contain an indicator whether it is a fraud transaction which is necessary to apply supervised machine learning methods. It should also contains data about the transactions, e.g. transaction details, customer details, etc., which could be used to predict the fraud transaction.
- The data contains target column 'Class' that indicates whether it is a fraud transaction. It also contains other columns ('Time', 'Amount', 'V1-V28') which can be used as factors to predict the target column 'Class'.
- There are some caveats for using the dataset
    - The dataset is a transformed dataset with PCA-transformed columns V1 - V28. Without the original dataset, we are not sure how the PCA transformation is performed. We cannot guarantee if the same transformation can be applied to any new data (e.g. real-world unseen data)
    - From interpretation perspective, we do not understand the meaning of the columns V1 - V28, which we fail to draw any meaningful conclusion betwen the factors and the target. Moreover, we might not detect any bias or unfairness among the ML model if the original dataset contains sensitive information, e.g. age.
- {dataset URL}

2. Clearly define the expected input and the 'ideal' output. Determine if machine learning is the appropriate method for addressing this problem.
- Data Input: Data about the transactions, e.g.
    - merchandise details: time, amount, merchant information, etc.
    - customer details: account balance, cardholder demographics, etc.
- Data Output: A soft prediction (probability) whether a transaction is a fraud
- As we can clearly define the expected input and output and we can collect the data, machine learning is the appropriate method to address the problem

3. If machine learning is deemed suitable, what should the model aim to achieve? How would you measure the model's performance?
- Objective: detect fraudulent transaction by a soft prediction (probability) and a threshold to determine whether a transaction is a fraud
- We can expect there would be potential false positives (clean transactions predicted as fraud) and false negatives (fraudulent transactions predicted a clean) from the prediction and the model should balance both based on our objectives. For example, more false negatives might create bad debts, while more false positives might affect the revenue.
- Moreover, the dataset is an unbalanced dataset with small amount of fraud records.
- We can use evaluation metrics, e.g. precision, recall, area under precision-recall curve, etc. to measure the model performance (e.g. how accurate the predictions are, how balanced between false positives and false negatives, etc.)


4. How would a human tackle this issue? Can you propose any heuristic methods to solve this problem?
5. What are the major steps required to resolve this problem?

- It would be a data collection - analysis - decision making - remedy approach to solve this issue by human
    - data collection: collect the recent N days transactions and the related information, e.g. transaction, merchant, cardholder, etc.
    - analysis: based on personal experience and insights drawn from past identified fraudulent transactions, a group of fraud experts will try to identify if there is any pattern similar to the traits of fraud based on the collected data. They will then provide a score of fraud
    - decision making: a cutoff of the score will be made to determine if a transaction is classified as a fraud
    - remedy: If the transaction is clean, no action is required. If it is a fraud, the money transfer will be on-hold. The customer and the merchant will be contacted for further investigation.

6. Draw a diagram that illustrates the input, output, and key stages of the problem-solving process.

7. Which type of machine learning would be best suited for your problem?
8. What specific machine learning technique would be most effective for this problem?
- Given a clear target with labelled data, supervised machine learning will be best suited.
- Simple linear regression can be first applied to observe any linear relationship between the target and the factors and obtain a baseline performance.
- Given the possibility of non-linear relationship, non-linear algorithm, e.g. SVM, can be applied to see if any improvement can be obseved.
- Given the variety of data types (textual, numerical, etc.), decision tree based algorithm can be applied to see if any improvement can be obseved.

**Comment by kvarada** [comment]
@tonyshumlh 

Thanks for putting this together. I'm not sure whether it's ready for my review yet but I would also describe a scenario and your objective. Something as shown below:  

> Imagine you work at a bank, and the current fraud detection model isn't performing well. Your boss asks you to explore machine learning approaches to improve the detection and flagging of fraudulent credit card transactions. While researching online, you find [Credit Card Fraud Detection dataset on Kaggle](https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud) that could be useful for creating a prototype.

And then ask them to brainstorm the provided questions

**Comment by tonyshumlh** [comment]
@kvarada I agree. I have refined the problem framing example in `workshop-12.qmd` and added your questions in `workshop-10.qmd`
I would continue to consolidate the example topics + dataset in `workshop-11.qmd`

**Comment by tonyshumlh** [comment]
@kvarada @p-bajpai @p-bajpai Added the example topics and datasets. Great if you can review the content and the layout (there is some issue with my Quarto preview). Thank you.

File: website/workshop-11.qmd [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+::: {.callout-tip collapse="true"}
+#### Stock Market Sentiment
+
+Imagine you work at a bank, and the current workflow to understand the stock market sentiment isn't performing well. Your boss asks you to explore machine learning approaches to improve the detection of stock market sentiment, which is cruical to some of the trading strategies, based on comments in social media platform. While researching online, you find [this dataset on Kaggle](https://www.kaggle.com/datasets/utkarshxy/stock-markettweets-lexicon-data) that could be useful for creating a prototype.
+:::
+
+::: {.callout-tip collapse="true"}
+#### E-commerce Strategy
+
+Imagine you are a consultant hired by an E-commerce platform. You are asked to analyze the retail data and explore machine learning approaches to address 1 of the planned initiatives of the company, which includes but not limited to:
```
Comment (kvarada) [inline website/workshop-11.qmd:26]: 1 of the planned initiatives --> one of the planned initiatives

File: website/workshop-12.qmd [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+## Scenario
+
+Imagine you work at a bank, and the current fraud detection workflow isn't performing well. Your boss asks you to explore machine learning approaches to improve the detection and flagging of fraudulent credit card transactions. While researching online, you find [Credit Card Fraud Detection dataset on Kaggle](https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud) that could be useful for creating a prototype.
+
+## Problem Framing
+
+#### Is the provided dataset appropriate for the specified objective? What type of data would ideally solve your problem or research question? Are there better-suited datasets available on the internet for this objective?
+
+-   The ideal dataset should contain an indicator whether it is a fraud transaction, which is necessary for applying supervised machine learning methods. It should also contains data about the transactions, e.g. transaction details, customer details, etc., which could be used to classify the fraud transaction.
```
Comment (kvarada) [inline website/workshop-12.qmd:19]: Here you're assuming that supervised ML is the ideal approach? You could also treat it as an outlier detection problem.

**Review by kvarada — CHANGES_REQUESTED** [review]
Overall, it's looking great! Thank you for doing this @tonyshumlh . I'm requesting some minor changes.

File: website/workshop-12.qmd (hunk 2 above)
Comment (tonyshumlh) [reply website/workshop-12.qmd:19]: If we have labelled data, supervised ML should be a straight-forward approach. If the data is not labelled, outlier detection should be great to identify transactions deviated much from the normal one.

**Pull request merged**